app/app_main.c

```diff
@@ -3719,6 +3765,114 @@ static ACVP_RESULT app_kas_ecc_handler(ACVP_TEST_CASE *test_case)
     return rv;
 }
 
+static ACVP_RESULT app_kas_ffc_handler(ACVP_TEST_CASE *test_case)
+{
+    ACVP_KAS_FFC_TC         *tc;
+    const EVP_MD *md = NULL;
+    ACVP_RESULT rv = ACVP_CRYPTO_MODULE_FAIL;
+    unsigned char *Z;
+    int Zlen;
+    DH *dh = NULL;
+    BIGNUM *peerkey = NULL;
+
+    tc = test_case->tc.kas_ffc;
+
+    switch (tc->md)
+    {
+    case ACVP_SHA224:
+        md = EVP_sha224();
+        break;
+    case ACVP_SHA256:
+        md = EVP_sha256();
+        break;
+    case ACVP_SHA384:
+        md = EVP_sha384();
+        break;
+    case ACVP_SHA512:
+        md = EVP_sha512();
+        break;
+    default:
+        printf("No valid hash name %d\n", tc->md);
+        return rv;
+        break;
+    }
+
+    dh = FIPS_dh_new();
+    if (!dh) {
+        return rv;
+    }
+    if (!BN_hex2bn(&dh->p, (char *)tc->p)) {
+            printf("BN_hex2bn failed p\n");
+            goto error;
+    }
+    if (!BN_hex2bn(&dh->q, (char *)tc->q)) {
+            printf("BN_hex2bn failed q\n");
+            goto error;
+    }
+
+    if (!BN_hex2bn(&dh->g, (char *)tc->g)) {
+            printf("BN_hex2bn failed g\n");
+            goto error;
+    }
+
+    if (!BN_hex2bn(&peerkey, (char *)tc->eps)) {
+            printf("BN_hex2bn failed eps\n");
+            goto error;
+    }
+
+    if (tc->test_type == ACVP_KAS_FFC_TT_VAL) {
+        if (!BN_hex2bn(&dh->priv_key, (char *)tc->epri)) {
+            printf("BN_hex2bn failed epri\n");
+            goto error;
+        }
+
+        if (!BN_hex2bn(&dh->pub_key, (char *)tc->epui)) {
+            printf("BN_hex2bn failed epui\n");
+            goto error;
+        }
+    }
+
+    if (tc->test_type == ACVP_KAS_FFC_TT_AFT) {
+        if (!DH_generate_key(dh)) {
+            printf("DH_generate_key failed\n");
+            goto error;
+        }
+    }
+    Z = OPENSSL_malloc(BN_num_bytes(dh->p));
+    if (!Z) {
+        printf("Malloc failed for Z\n");
+        goto error;
+    }
+
+    Zlen = DH_compute_key_padded(Z, peerkey, dh);
+    FIPS_digest(Z, Zlen, (unsigned char *)tc->chash, NULL, md);
+    tc->chashlen = M_EVP_MD_size(md);
+
+    if (tc->test_type == ACVP_KAS_FFC_TT_AFT) {
+        memcpy(tc->z, Z, Zlen);
+        tc->zlen = Zlen;
+    }
+
+    tc->piut = BN_bn2hex(dh->pub_key);
+    tc->piutlen = strnlen(tc->piut, ACVP_KAS_FFC_MAX_STR);
+
+
+    rv = ACVP_SUCCESS;
+
+error:
+    if (Z) {
+        OPENSSL_cleanse(Z, Zlen);
+    }
+    OPENSSL_free(Z);
+    BN_clear_free(peerkey);
+    BN_clear_free(dh->priv_key);
+    BN_clear_free(dh->pub_key);
+    dh->priv_key = NULL;
+    dh->pub_key = NULL;
+	FIPS_dh_free(dh);
```

Comment:
Nit, too much space

Reply:
yep, one of those nasty tabs got left in there.